include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -0,0 +1,616 @@
+    }
+
+    cuvm::CUAddressRange    address_range;
+    vector<mem_handle_t>    mapping;
+    bitmask                 mapped, available;
+    ptrdiff_t               available_blocks = 0;
+    size_t                  block_size = 0;
+  };
+
+  SmallVector<va_region, 8> va_regions_;
+  SmallVector<cuvm::CUMemAddressRange, 8> va_ranges_;
+  size_t initial_va_size_ = 0;
+  size_t block_size_ = 0;
+  size_t total_mem_ = 0;
+  int device_ordinal_ = -1;
+
+  void adjust_params(size_t &size, size_t &alignment) {
+    alignment = std::max(alignment, next_pow2(size >> 11));
+    alignment = std::min(alignment, block_size_);
+    size = align_up(size, alignment);
+  }
+
+  void *do_allocate(size_t size, size_t alignment) override {
+    adjust_params(size, alignment);
+    if (size == 0)
+      return nullptr;
+    lock_guard pool_guard(pool_lock_);
+    // try to get a free region that's already mapped
+    void *ptr = try_get_mapped(size, alignment);
+    if (ptr) {
+      stat_add_allocation(size);
+      return ptr;
+    }
+    DeviceGuard dg(device_ordinal_);
+    mem_lock_guard mem_guard(mem_lock_);
+    ptr = get_va(size, alignment);
+    map_storage(ptr, size);
+    ptr = try_get_mapped(size, alignment);
+    assert(ptr != nullptr);
+    stat_add_allocation(size);
+    return ptr;
+  }
+
+  void *try_get_mapped(size_t size, size_t alignment) {
+    char *ptr = static_cast<char*>(free_mapped_.get(size, alignment));
+    if (ptr) {
+      stat_take_free(size);
+      free_va_.get_specific_block(ptr, ptr + size);
+      mark_as_unavailable(ptr, size);
+      return ptr;
+    } else {
+      return nullptr;
+    }
+  }
+
+  void *get_va(size_t size, size_t alignment) {
+    void *ptr = free_va_.get(size, alignment);
+    if (ptr)
+      return ptr;
+    va_allocate(size);
+    ptr = free_va_.get(size, alignment);
+    assert(ptr);
+    return ptr;
+  }
+
+  /**
+   * @brief Allocate virtual address space of at at least `min_size` bytes.
+   *
+   * The function hints the driver to use a distinct address space for each device in the
+   * attempt to have a contiguous address spaces for each device. Currently, the spacing
+   * between device VA spaces is 1 TiB and initial VA size for each device is 4 GiB.
+   */
+  void va_allocate(size_t min_size) {
+    size_t va_size = std::max(next_pow2(min_size), initial_va_size_);
+
+    CUdeviceptr hint = {};
+    va_region *last_va = nullptr;
+    if (va_regions_.empty()) {
+      // Calculate the hint address for the allocations for this device.
+      // There's some not very significant base and different devices get
+      // address spaces separated by 2^40 - this should be quite enough.
+      hint = (device_ordinal_ + 1) * 0x10000000000u;
+    } else {
+      // Try to allocate after the last VA for this device
+      last_va = &va_regions_.back();
+      hint = last_va->address_range.end();
+    }
+    va_ranges_.push_back(cuvm::CUMemAddressRange::Reserve(va_size, 0, hint));
```

Comment:
Can `Reserve` fail? What then?

Reply:
Note that we haven't change anything in the resource yet. If it fails, we simply quit with an exception (likely CUDABadAlloc) which is the expected behavior. It can also fail with `CUDA_ERROR_NOT_SUPPORTED` - it's also OK, since the user should have checked the support for VMM API before using this resource.

Reply:
Throw is good enough.